src/devtools/devtoolsAtom.ts

```diff
@@ -0,0 +1,133 @@
+}
+
+const devtoolAtomCache = new WeakMap()
+
+// TODO: this was an atom family, but i've changed it to a fn
+// this will probably result in an infinite loop when used unless memoized
+export const devtoolAtom = (anAtom: PrimitiveAtom<any>) =>
+  atom((get) => {
+    if (!IS_DEVTOOL_ENV) return
+    const isWriteable = !!anAtom.write
+    const isTimeTravelingAtom = atom(false)
+    const lastValueAtom = atom(undefined)
+    const atomName = anAtom.debugLabel || anAtom.toString()
+    const extension = getExtension()
+    const cachedDevtools = devtoolAtomCache.get(anAtom)
+    const devtools: Devtools =
+      cachedDevtools || extension.connect({ name: atomName })
+
+    let unsubscribe: undefined | (() => void)
+
+    const subscribeAtom = atom<void, { type: 'mount' }>(
+      (get) => {
+        if (isWriteable) {
+          const isTimeTraveling = get(isTimeTravelingAtom)
+          if (!isTimeTraveling)
+            devtools.send(
+              `${atomName} - ${new Date().toLocaleString()}`, // TODO: maybe someone would want to customize this
+              get(anAtom)
+            )
+        }
+      },
+      (get, set, update) => {
+        const listener = (message: Message) => {
+          if (message.type === 'DISPATCH' && message.state) {
+            if (
+              message.payload?.type === 'JUMP_TO_ACTION' ||
+              message.payload?.type === 'JUMP_TO_STATE'
+            ) {
+              set(isTimeTravelingAtom, true) // TODO: when to set false?
+            }
+            if (message.payload?.type !== 'TOGGLE_ACTION')
+              set(anAtom, JSON.parse(message.state))
+          } else if (
+            message.type === 'DISPATCH' &&
+            message.payload?.type === 'COMMIT'
+          ) {
+            devtools.init(get(lastValueAtom))
+          } else if (
+            message.type === 'DISPATCH' &&
+            message.payload?.type === 'IMPORT_STATE'
+          ) {
+            const computedStates =
+              message.payload.nextLiftedState?.computedStates || []
+            computedStates.forEach(
+              ({ state }: { state: any }, index: number) => {
+                if (index === 0) {
+                  devtools?.init(state)
+                } else {
+                  set(anAtom, state)
+                }
+              }
+            )
+          }
+        }
+
+        switch (update.type) {
+          case 'mount': {
+            unsubscribe = isWriteable ? devtools.subscribe(listener) : undefined
+            if (devtools.shouldInit) {
+              devtools.init(get(anAtom))
+              devtools.shouldInit = false
+              devtoolAtomCache.set(anAtom, devtools)
+            }
+          }
+        }
+      }
+    )
+
+    subscribeAtom.onMount = (setAtom) => {
+      setAtom({ type: 'mount' })
+      return () => unsubscribe?.()
+    }
+
+    get(subscribeAtom)
+
+    if (!isWriteable) {
+      devtools.send(`${atomName} - ${new Date().toLocaleString()}`, get(anAtom))
```

Comment:
this is a side effect in `read`.
conceptually `read` must be pure, and it may be double invoked in the future or render without commit (if async somewhere).

I think it would be nice to send in `write` which solved double invocation, but render without commit can still happen. If we want to see something committed, `useAtomDevtools` would be better.